ant/src/plus/profiles/speed_and_cadence/mod.rs

```diff
@@ -0,0 +1,119 @@
+use derive_new::new;
+use packed_struct::derive::PackedStruct;
+
+pub const DEVICE_TYPE: u8 = 121;
+
+#[derive(PackedStruct, new, PartialEq, Copy, Clone, Debug)]
+#[packed_struct(endian = "lsb")]
+pub struct SpeedAndCadence {
+    /// Time of the last valid bike cadence event (1/1024 sec)
+    pub cadence_event_time: u16,
+
+    /// Total number of pedal revolutions
+    pub cadence_revolution_count: u16,
+
+    /// Time of the last valid bike speed event (1/1024 sec)
+    pub speed_event_time: u16,
+
+    /// Total number of wheel revolutions
+    pub speed_revolution_count: u16,
+}
+
+impl SpeedAndCadence {
+    /// Calculates the average cadence (rpm)
+    pub fn cadence(a: SpeedAndCadence, b: SpeedAndCadence) -> Option<f32> {
```

Comment:
> Given you are implementing this on a struct, you could switch the first arg to `&self` so you only have to pass in the second struct for the delta.

I guess my only gripe with that is I'm not sure if this should be a method on the first or the 2nd message. All the calculations require 2 messages, and calculate data between those two points in time.

When you do the calculation, you want to see the value at the time the 2nd message was received. So would the method be on the 2nd message and require a `previous: SpeedAndCadence` or would you call the method on the first, and specify the next.
I couldn't decide, so I just made it this way

Reply:
I think that is a matter of style/preference. Either is fine. To be honest, you could do both and just make them call this function in behind.

Also another thing, you currently have them consuming the type, is that intended or do you want to just pass a reference?

Reply:
@cujomalainey I have the `Copy` trait implemented. Afaik that will copy it instead of taking ownership. Am I missing something here?

Reply:
> @cujomalainey I have the `Copy` trait implemented. Afaik that will copy it instead of taking ownership. Am I missing something here?

Ah I missed the copy, yep that is fine, but you are copying the struct unnecessarily, its only 8 bytes, but you can save that by doing a reference.

Reason I am being a bit nitpicky is I am trying to target embedded devices eventually.

Reply:
I'm happy to change anything on this, and appreciate the feedback.

I would also like to run this on embedded at some point too. Copy seemed like the way to go since this is only 8 bytes. Its the same size as an int64. I wouldn't pass a ref to that 🤷 .

How would you decide between copy vs ref?

Reply:
I typically use 2 measures.

One is modification, just in case the type size changes out from under me I'll pass a reference.

Second on embedded it's usually 32bit (or smaller) sized registers so anything larger than that I'll typically send via ref.

Lastly I just like the style because it forces the type to be RO in the function and a future modification cant make it RW accidentally and then wonder why the change disappeared (for all the new rust users)